### PR TITLE
8321940: Improve CDSHeapVerifier in handling of interned strings

### PR DESCRIPTION
**Pull request by iklam**
The PR fixes the CDSHeapVerifier so that it no longer warns about interned strings that are not referenced by any of the archived heap object graphs. Please see the bug report for more details.

Tests: tiers 1-4, builds-tier5

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8321940](https://bugs.openjdk.org/browse/JDK-8321940): Improve CDSHeapVerifier in handling of interned strings (**Enhancement** - P4)

### Reviewers
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to [842ad51c](https://git.openjdk.org/jdk/pull/17102/files/842ad51cf209eb13b551b70117d8caf2c1e7bd8d)
 * [Matias Saavedra Silva](https://openjdk.org/census#matsaave) (@matias9927 - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17102/head:pull/17102` \
`$ git checkout pull/17102`

Update a local copy of the PR: \
`$ git checkout pull/17102` \
`$ git pull https://git.openjdk.org/jdk.git pull/17102/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17102`

View PR using the GUI difftool: \
`$ git pr show -t 17102`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17102.diff">https://git.openjdk.org/jdk/pull/17102.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17102#issuecomment-1855216324)

### CODE REVIEW

**Review by calvinccheung — COMMENTED** [review]
With your new changes, item C in the comment can be removed.

```
80 // [C] A non-final static string that is assigned a string literal during class
 81 //     initialization; this string is never changed during -Xshare:dump.
```

**Comment by iklam** [comment]
> With your new changes, item C in the comment can be removed.
> 
> ```
>  80 // [C] A non-final static string that is assigned a string literal during class
>  81 //     initialization; this string is never changed during -Xshare:dump.
> ```

Hi Calvin:

Thanks for the review. [C] can still be useful for checking problematic code like this (a variation of the example in cdsHeapVerifier.cpp)

We want cdsHeapVerifier to flag such patterns. However, if we know that `someCondition()` always returns the same condition, then we can add an exclusion of type [C] so cdsHeapVerifier won't complain about it.

```
class Foo {
    final Foo archivedFoo; // this field is archived by CDS
    String bar;
    static {
        CDS.initializeFromArchive(Foo.class);
        if (archivedFoo == null) {
            archivedFoo = new Foo();
            archivedFoo.bar = Bar.bar;
        }
    }
    static Foo get() { return archivedFoo; }
    boolean test() {
        return bar == Bar.bar;
    }
}

class Bar {
    // this field is initialized in both CDS dump time and runtime.
    // It may hold a different value depending on someCondition()
    static String bar = (someCondition()) ? "xxx" : "yyy";
}
```

**Comment by calvinccheung** [comment]
Ok. Thanks for the explanation.

**Review by calvinccheung — APPROVED** [review]

File: src/hotspot/share/cds/cdsHeapVerifier.cpp [hunk 1]
```diff
@@ -245,6 +215,12 @@ inline bool CDSHeapVerifier::do_entry(oop& orig_obj, HeapShared::CachedOopInfo&
 
   StaticFieldInfo* info = _table.get(orig_obj);
   if (info != nullptr) {
+    if (value.orig_referrer() == nullptr && java_lang_String::is_instance(orig_obj)) {
+      // This string object is not refernced by any of the archived object graphs. It's archived
```
Comment (dholmes-ora) [inline src/hotspot/share/cds/cdsHeapVerifier.cpp:219]: typo: refernced

**Review by matias9927 — APPROVED** [review]
Looks good! The comments do a good job explaining the purpose of this code.

**Comment by iklam** [comment]
Thanks @calvinccheung and @matias9927 for the review.
/integrate

**Pull request closed without merging**